metagraph/src/graph/representation/succinct/dbg_succinct_cached.hpp

```diff
@@ -0,0 +1,235 @@
+#ifndef __DBG_SUCCINCT_CACHED__
+#define __DBG_SUCCINCT_CACHED__
+
+#include <cache.hpp>
+#include <lru_cache_policy.hpp>
+
+#include "boss.hpp"
+#include "dbg_succinct.hpp"
+#include "common/seq_tools/reverse_complement.hpp"
+
+namespace mtg {
+namespace graph {
+
+class DBGSuccinctCached : public DeBruijnGraph {
+  public:
+    typedef boss::BOSS::edge_index edge_index;
+    typedef boss::BOSS::TAlphabet TAlphabet;
+
+    template <typename Key, typename Value>
+    using LRUCache = caches::fixed_sized_cache<Key, Value, caches::LRUCachePolicy<Key>>;
+
+    typedef LRUCache<node_index, std::string> DecodedCache;
+    typedef LRUCache<edge_index, edge_index> BwdFirstCache;
+
+    DBGSuccinctCached(const DBGSuccinct &dbg, size_t cache_size = 1000)
+          : graph_(dbg), boss_(graph_.get_boss()),
+            bwd_first_cache_(cache_size),
+            decoded_cache_(cache_size * dbg.alphabet().size()) {}
```

Comment:
Would be good to add benchmarks similar to https://github.com/ratschlab/metagraph/blob/master/metagraph/benchmarks/graph/bench_boss.cpp

So then we can work on improving this implementation more easily and get more confident improvements.